File: lib/services/audio_engine.dart
import 'dart:async';
import 'dart:io';
import 'package:audioplayers/audioplayers.dart';
import 'package:flutter_tts/flutter_tts.dart';

import 'windows_tts.dart';

/// Multi-track engine. See DESIGN.md §4 for behavior.
///
/// Tracks:
///   buzz     — one-shot focus reminder
///   pullover — looped pull-over voice
///   siren    — looped critical siren (with volume ducking)
///   dialer   — one-shot 3-tone dialing for "112" (offsets at 0.071, 0.437, 0.701 s)
///   calling  — ringback played exactly 3 times then stops
class AudioEngine {
  final AudioPlayer _buzz = AudioPlayer();
  final AudioPlayer _pullover = AudioPlayer();
  final AudioPlayer _siren = AudioPlayer();
  final AudioPlayer _dialer = AudioPlayer();
  final AudioPlayer _calling = AudioPlayer();
  // Dispatcher pickup ("911, what's your emergency?") — plays once after the
  // ringback ends. Siren stays ducked until the whole dispatcher chain finishes.
  final AudioPlayer _accept = AudioPlayer();
  StreamSubscription<void>? _acceptSub;

  // Pre-recorded fixed prefix: "A driver has become unresponsive. Please send
  // help to ...". Generated externally (TTSforge) and shipped as an asset so we
  // don't pay live-TTS latency for this every emergency.
  final AudioPlayer _intro = AudioPlayer();
  StreamSubscription<void>? _introSub;

  // Live TTS for the variable location tail (Windows uses native SAPI).
  final FlutterTts _tts = FlutterTts();
  String? _emergencyLocationText;
  bool _ttsConfigured = false;

  static const dialDigitOffsetsMs = <int>[71, 437, 701];

  double _master = 1.0;
  int _callingPlaysLeft = 0;
  StreamSubscription<void>? _callingSub;

  Future<void> init() async {
    await _pullover.setReleaseMode(ReleaseMode.loop);
    await _siren.setReleaseMode(ReleaseMode.loop);
    await _buzz.setPlayerMode(PlayerMode.lowLatency);

    _callingSub = _calling.onPlayerComplete.listen((_) async {
      _callingPlaysLeft -= 1;
      if (_callingPlaysLeft > 0) {
        await _calling.play(AssetSource('sounds/calling.mp3'),
            volume: _master);
      } else {
        // Ringback finished → play the dispatcher pickup (911 accept).
        // Siren stays ducked across the full dispatcher chain.
        await _accept.play(AssetSource('sounds/911accept.mp3'),
            volume: _master);
      }
    });

    // After the dispatcher's "911 what's your emergency" → play the cached
    // prefix ("A driver has become unresponsive. Please send help to").
    _acceptSub = _accept.onPlayerComplete.listen((_) async {
      await _intro.play(AssetSource('sounds/dispatch_intro.mp3'),
          volume: _master);
    });

    // After the cached prefix → speak the location tail live, then ramp the
    // siren back up once TTS is done.
    _introSub = _intro.onPlayerComplete.listen((_) async {
      await _speakLocation();
      await _rampVolume(_siren, _master, 200);
    });

    await _configureTts();
  }

  Future<void> _configureTts() async {
    if (_ttsConfigured) return;
    try {
      await _tts.setLanguage('en-US');
      await _tts.setSpeechRate(0.45); // slower & clearer
      await _tts.setPitch(1.0);
      await _tts.setVolume(1.0);
      // Try to pick a US female voice. Voice catalogue varies per platform —
      // we look for one that contains "female" or the typical Windows SAPI
      // names ("Zira" / "Aria"). If none match, we accept the default.
      final voices = await _tts.getVoices;
      if (voices is List) {
        for (final v in voices) {
          if (v is Map) {
            final name = (v['name'] ?? '').toString().toLowerCase();
            final locale = (v['locale'] ?? '').toString().toLowerCase();
            final isUs = locale.startsWith('en-us') || locale.startsWith('en_us');
            final isFemale = name.contains('female') ||
                name.contains('zira') ||
                name.contains('aria') ||
                name.contains('jenny');
            if (isUs && isFemale) {
              await _tts.setVoice({
                'name': v['name'].toString(),
                'locale': v['locale'].toString(),
              });
              break;
            }
          }
        }
      }
      _ttsConfigured = true;
    } catch (_) {
      // best-effort; default voice is fine if any setter fails
    }
  }

  /// Set by the alert engine (or DrivePage) before the emergency flow starts.
  /// If null when the chain reaches TTS, we speak a generic fallback.
  void setEmergencyLocationText(String? text) {
    _emergencyLocationText = text;
  }

  // Dedicated player for the boosted-gain TTS WAV on Windows.
  final AudioPlayer _ttsPlayer = AudioPlayer();

  Future<void> _speakLocation() async {
    final text = _emergencyLocationText ?? 'an unknown location';

    // On Windows, render TTS to a WAV via SAPI then amplify so the dispatcher
    // location matches the volume of the pre-recorded prefix. flutter_tts
    // can't boost above 1.0 on its own.
    if (Platform.isWindows) {
      final path = await WindowsTts.synthesize(text: text, gainDb: 8.0);
      if (path != null) {
        final completer = Completer<void>();
        late StreamSubscription<void> sub;
        sub = _ttsPlayer.onPlayerComplete.listen((_) {
          sub.cancel();
          if (!completer.isCompleted) completer.complete();
        });
        try {
          await _ttsPlayer.play(DeviceFileSource(path), volume: _master);
        } catch (_) {
          if (!completer.isCompleted) completer.complete();
        }
        return completer.future.timeout(const Duration(seconds: 15),
            onTimeout: () {});
      }
      // PowerShell synth failed → fall through to flutter_tts.
    }

    final completer = Completer<void>();
    _tts.setCompletionHandler(() {
      if (!completer.isCompleted) completer.complete();
    });
    _tts.setErrorHandler((_) {
      if (!completer.isCompleted) completer.complete();
    });
    try {
      await _tts.awaitSpeakCompletion(true);
      await _tts.speak(text);
      if (!completer.isCompleted) completer.complete();
    } catch (_) {
      if (!completer.isCompleted) completer.complete();
    }
    return completer.future.timeout(const Duration(seconds: 12),
        onTimeout: () {});
  }

  Future<void> dispose() async {
    await _callingSub?.cancel();
    await _acceptSub?.cancel();
    await _introSub?.cancel();
    await _tts.stop();
    await _buzz.dispose();
    await _pullover.dispose();
    await _siren.dispose();
    await _dialer.dispose();
    await _calling.dispose();
    await _accept.dispose();
    await _intro.dispose();
    await _ttsPlayer.dispose();
  }

  void setMasterVolume(double v) {
    _master = v.clamp(0, 1);
  }

  // ----- public ops -----
  Future<void> playBuzz() async {
    await _buzz.stop();
    await _buzz.play(AssetSource('sounds/buzz.mp3'), volume: _master);
  }

  Future<void> startPullover() async {
    if (_pullover.state == PlayerState.playing) return;
    // Re-assert loop each start: audioplayers can drop the release mode after
    // a stop()/play() cycle on Windows.
    await _pullover.setReleaseMode(ReleaseMode.loop);
    await _pullover.play(AssetSource('sounds/PULLOVER.mp3'), volume: _master);
  }

  Future<void> stopPullover() => _pullover.stop();

  Future<void> startSiren() async {
    if (_siren.state == PlayerState.playing) return;
    await _siren.setReleaseMode(ReleaseMode.loop);
    await _siren.setVolume(_master);
    await _siren.play(AssetSource('sounds/sirenLoop.mp3'), volume: _master);
  }

  Future<void> stopSiren() => _siren.stop();
  // Aggressive duck during the dispatcher chain so the operator + TTS are
  // clearly audible over the alarm.
  Future<void> duckSiren() => _rampVolume(_siren, _master * 0.10, 200);
  Future<void> unduckSiren() => _rampVolume(_siren, _master, 200);

  /// Plays the dialer one-shot. Returns the digit-offset timeline (ms) so the UI
  /// can light up digits in sync.
  Future<List<int>> playDialer() async {
    await _dialer.stop();
    await _dialer.play(AssetSource('sounds/dialingButtons.m4a'),
        volume: _master);
    return dialDigitOffsetsMs;
  }

  Future<void> onDialerEnd(void Function() cb) async {
    late StreamSubscription<void> sub;
    sub = _dialer.onPlayerComplete.listen((_) {
      sub.cancel();
      cb();
    });
  }

  Future<void> playCallingTimes(int n) async {
    _callingPlaysLeft = n;
    await _calling.play(AssetSource('sounds/calling.mp3'), volume: _master);
  }

  Future<void> stopCalling() async {
    _callingPlaysLeft = 0;
    await _calling.stop();
  }

  Future<void> stopAll() async {
    await Future.wait([
      stopPullover(),
      stopSiren(),
      stopCalling(),
      _dialer.stop(),
      _accept.stop(),
      _intro.stop(),
      _tts.stop(),
    ]);
  }

  Future<void> _rampVolume(AudioPlayer p, double target, int ms) async {
    // audioplayers has no built-in fade; do it in N steps.
    const steps = 10;
    final stepMs = (ms / steps).round();
    final start = await _readVolume(p);
    for (var i = 1; i <= steps; i++) {
      final v = start + (target - start) * (i / steps);
      await p.setVolume(v.clamp(0, 1));
      await Future.delayed(Duration(milliseconds: stepMs));
    }
  }

  // audioplayers doesn't expose current volume; store last set externally.
  // We approximate by ramping from `_master` for the unduck case and
  // from `_master * 0.25` for duck.
  Future<double> _readVolume(AudioPlayer p) async {
    if (identical(p, _siren)) return _master; // best-effort
    return _master;
  }
}

File: webapp/src/lib/audioEngine.ts
// Browser port of lib/services/audio_engine.dart.
//
// Multi-track engine built on HTMLAudioElement. See DESIGN.md §4.
//   buzz     — one-shot focus reminder
//   pullover — looped pull-over voice
//   siren    — looped critical siren (volume ducking during dispatcher chain)
//   dialer   — one-shot 3-tone "112" dialing (digit offsets 71/437/701 ms)
//   calling  — ringback, played exactly 3 times
//   accept   — "911, what's your emergency"
//   intro    — pre-recorded dispatcher prefix
//   then live TTS speaks the location, and the siren ramps back up.

import { speak } from "./tts";
import { dlog } from "./debug";

const SND = "/sounds";

function shortSrc(src: string): string {
  return src.split("/").pop() ?? src;
}
export const DIAL_DIGIT_OFFSETS_MS = [71, 437, 701];

function mkAudio(src: string, loop = false): HTMLAudioElement {
  const a = new Audio(src);
  a.loop = loop;
  a.preload = "auto";
  return a;
}

export class AudioEngine {
  private buzz = mkAudio(`${SND}/buzz.mp3`);
  private pullover = mkAudio(`${SND}/PULLOVER.mp3`, true);
  private siren = mkAudio(`${SND}/sirenLoop.mp3`, true);
  private dialer = mkAudio(`${SND}/dialingButtons.m4a`);
  private calling = mkAudio(`${SND}/calling.mp3`);
  private accept = mkAudio(`${SND}/911accept.mp3`);
  private intro = mkAudio(`${SND}/dispatch_intro.mp3`);

  private master = 1.0;
  private callingPlaysLeft = 0;
  // True only while a real emergency dispatcher chain (calling → accept → intro
  // → TTS) is in progress. The chained "ended" handlers below must NOT fire
  // otherwise — e.g. during unlock(), where every clip is played briefly to
  // unlock it; without this guard, the calling clip's "ended" would spuriously
  // launch the 911-accept audio.
  private dispatcherActive = false;
  private emergencyLocationText: string | null = null;
  private dialerEndCb: (() => void) | null = null;
  private rampTimer: ReturnType<typeof setInterval> | null = null;

  init() {
    // Ringback chains: replay up to 3 times, then dispatcher pickup.
    this.calling.addEventListener("ended", () => {
      dlog(`chain: calling ended (active=${this.dispatcherActive}, left=${this.callingPlaysLeft})`);
      if (!this.dispatcherActive) return;
      this.callingPlaysLeft -= 1;
      if (this.callingPlaysLeft > 0) {
        void this.play(this.calling, this.master);
      } else {
        void this.play(this.accept, this.master);
      }
    });
    // After dispatcher pickup → cached prefix.
    this.accept.addEventListener("ended", () => {
      dlog(`chain: accept ended (active=${this.dispatcherActive}) → intro`);
      if (!this.dispatcherActive) return;
      void this.play(this.intro, this.master);
    });
    // After prefix → speak the location, then bring the siren back.
    this.intro.addEventListener("ended", async () => {
      dlog(`chain: intro ended (active=${this.dispatcherActive}) → speak location="${this.emergencyLocationText ?? "(none)"}"`);
      if (!this.dispatcherActive) return;
      await speak(this.emergencyLocationText ?? "an unknown location", this.master);
      this.unduckSiren();
      this.dispatcherActive = false;
    });
    // Dialer one-shot end.
    this.dialer.addEventListener("ended", () => {
      dlog("chain: dialer ended → callback");
      const cb = this.dialerEndCb;
      this.dialerEndCb = null;
      cb?.();
    });
  }

  /** Browsers require a user gesture to unlock audio. Call from a click. */
  async unlock() {
    const all = [this.buzz, this.pullover, this.siren, this.dialer,
      this.calling, this.accept, this.intro];
    // Pass 1: start every clip muted within the user gesture to unlock it.
    for (const a of all) {
      a.muted = true;
      // Don't let the looping siren/pullover run away if a pause gets dropped
      // (iOS); startSiren()/startPullover() re-enable looping when really used.
      a.loop = false;
      try {
        await a.play();
      } catch {
        /* autoplay blocked / interrupted — ignore */
      }
    }
    // Pass 2: hard-stop everything. Pausing here, after all the play() promises
    // have settled, is reliable on iOS — where pause-immediately-after-play is
    // frequently ignored, which used to leak the siren + dispatcher audio.
    for (const a of all) {
      a.pause();
      a.currentTime = 0;
      a.muted = false;
    }
  }

  setMasterVolume(v: number) {
    this.master = Math.max(0, Math.min(1, v));
  }

  setEmergencyLocationText(text: string | null) {
    this.emergencyLocationText = text;
  }

  private async play(a: HTMLAudioElement, volume: number) {
    try {
      a.volume = Math.max(0, Math.min(1, volume));
      a.currentTime = 0;
      await a.play();
    } catch (e) {
      dlog(`audio play FAILED: ${shortSrc(a.src)} — ${(e as Error).name}: ${(e as Error).message}`);
    }
  }

  async playBuzz() {
    this.buzz.pause();
    await this.play(this.buzz, this.master);
  }

  async startPullover() {
    if (!this.pullover.paused) return;
    this.pullover.loop = true;
    await this.play(this.pullover, this.master);
  }
  stopPullover() {
    this.pullover.pause();
    this.pullover.currentTime = 0;
  }

  async startSiren() {
    if (!this.siren.paused) return;
    this.siren.loop = true;
    await this.play(this.siren, this.master);
  }
  stopSiren() {
    this.siren.pause();
    this.siren.currentTime = 0;
  }
  // iOS Safari ignores HTMLAudioElement.volume (it's read-only on iPhone), so
  // volume-ducking is silent there and the dispatcher chain gets buried under
  // the siren. Pause the siren outright during the dispatcher chain instead —
  // reliable on every platform — then resume it once the location is spoken.
  duckSiren() {
    this.siren.pause();
  }
  unduckSiren() {
    if (this.siren.paused) {
      this.siren.loop = true;
      void this.play(this.siren, this.master);
    }
  }

  /** Plays the dialer one-shot. Returns the digit-offset timeline (ms). */
  async playDialer(): Promise<number[]> {
    dlog("chain: playDialer");
    await this.play(this.dialer, this.master);
    return DIAL_DIGIT_OFFSETS_MS;
  }

  onDialerEnd(cb: () => void) {
    this.dialerEndCb = cb;
  }

  async playCallingTimes(n: number) {
    dlog(`chain: playCallingTimes(${n})`);
    this.dispatcherActive = true; // arm the calling → accept → intro chain
    this.callingPlaysLeft = n;
    await this.play(this.calling, this.master);
  }

  stopCalling() {
    this.dispatcherActive = false;
    this.callingPlaysLeft = 0;
    this.calling.pause();
    this.calling.currentTime = 0;
  }

  stopAll() {
    this.dispatcherActive = false;
    if (this.rampTimer) {
      clearInterval(this.rampTimer);
      this.rampTimer = null;
    }
    this.stopPullover();
    this.stopSiren();
    this.stopCalling();
    for (const a of [this.dialer, this.accept, this.intro]) {
      a.pause();
      a.currentTime = 0;
    }
    if (typeof window !== "undefined" && "speechSynthesis" in window) {
      window.speechSynthesis.cancel();
    }
    this.dialerEndCb = null;
  }

  private rampVolume(a: HTMLAudioElement, target: number, ms: number) {
    if (this.rampTimer) clearInterval(this.rampTimer);
    const steps = 10;
    const stepMs = Math.round(ms / steps);
    const start = a.volume;
    let i = 0;
    this.rampTimer = setInterval(() => {
      i += 1;
      const v = start + (target - start) * (i / steps);
      a.volume = Math.max(0, Math.min(1, v));
      if (i >= steps && this.rampTimer) {
        clearInterval(this.rampTimer);
        this.rampTimer = null;
      }
    }, stepMs);
  }

  dispose() {
    this.stopAll();
  }
}
